### ISSUE TITLE
Cannot find java.exe within Windows Subsystem for Linux

### ISSUE THREAD

**Issue by niklaas** [ensime/ensime-vim#409]
I'm trying to set up a Scala development environment within Windows Subystem for Linux (WSL) using both JRE and JDK installed on Windows i.e., not within WSL. I exported `JAVA_HOME` and `JRE_HOME` to the appropriate folders and `alias java=java.exe`. This works quite well: `java -version` runs and `sbt` does its job too. I can populate an ensime configuration with `sbt ensimeConfig`.

The issue is that ensime-vim cannot find Java when starting vim. I suspect this is due to the fact that ensime-vim thinks it's on a Linux system and looks for `java` within `:java-home` (as set in `.ensime` config). However, since it's an installation on Windows, it's actually `java.exe`. (This is probably related to #383.)

Is there some way to configure ensime-vim to look for `java.exe` instead of `java`?

### Steps to reproduce

1. Set-up WSL to use JDK and JRE on Windows
2. Alias `java` to `java.exe`
3. Follow installation instructions of `ensime-vim`

### Expected behaviour

I open a `*.scala` file in vim on WSL and "everything just works".

### Actual behaviour

In vim ensime-vim comlains:

```
[ensime] Java not found or not executable, verify :java-home in your .ensime config
```

**Comment by ches** [comment]
Hi Niklaas, could you try checking what Python's notion of `os.name` is when running under WSL? In Vim you can try this:

```vim
:py import os; print(os.name)
```

Also check the value of `os.pathsep` too, we might want to update that part of #383's changes to try to rely more on Python's platform-agnosticism help too.

**Comment by niklaas** [comment]
Thanks for your prompt response.

`:py import os; print(os.name)` gives `posix`.

`:py import os; print(os.pathsep)` gives `:`.

For comparison: On gvim in Windows I get `nt` and `;` respectively.

**Comment by niklaas** [comment]
At the moment I don't have a Linux system at hand to check whether `posix` and `:` can be used to distinguish WSL from a normal Linux. Anyway, in my ZSH config I use `uname -a | grep Microsoft` to find out whether the shell is run within WSL or not:

```
> uname -a
Linux len-x1 4.4.0-43-Microsoft #1-Microsoft Wed Dec 31 14:42:53 PST 2014 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by ches** [comment]
> At the moment I don't have a Linux system at hand to check whether `posix` and `:` can be used to distinguish WSL from a normal Linux.

No, these values are the same as when running in Linux or OS X.

So, we could try checking for an executable named `java.exe` as a fallback even when `os.name` is `posix`, to handle this case of running in WSL when the Java installation comes from Windows. But the next question is how does `java.exe -cp` behave under these conditions? As a Windows executable is it necessary that the classpath entries given to it are separated by `;` even though it's invoked from WSL?

I don't have an environment suitable for testing this, so your quickest route to a solution will likely come if you can try to experiment yourself. The code touched by #383 should be the relevant areas you need to look at: finding the `java` executable and passing it a classpath (that code could be updated to use `os.pathsep.join(classpath)` instead of a manual ternary, aside from the unknown point in question).

I'd also like to understand the use cases—I'm not a Windows person at all, so I wonder about why you have (and want) this setup of JRE/JDK installed in Windows, and presumably Vim and maybe Python, etc. in WSL. Might life be easier in other ways too if you installed Java in WSL?
